### PR TITLE
Add Rowset push and shift

### PR DESCRIPTION
**Pull request by harmenjanssen**
For appending or prepending single rows to rowsets.

Same as `concat`, but when you have a single row. 

I've taken the names from `array_shift` and `array_push`, but now I'm doubting whether I shouldn't've just used `prepend` and `append`, what do you think?

### CODE REVIEW

**Comment by harmenjanssen** [comment]
Oh haha, I meant `unshift` indeed. 
Doesn't Laravel support "append" though? That's rather unsymmetrical... But `push` and `prepend` would be your preference for the functions in this PR, right?

**Comment by martijngastkemper** [comment]
Laravel Collection doesn't have a `append` method. That's done by `push` and `concat`. So I indeed suggest `push` and `prepend` to  make the switch to Laravel easier.

**Comment by harmenjanssen** [comment]
Makes sense! I've renamed the `shift` method.

**Comment by harmenjanssen** [comment]
I've also updated `map` and `filter` in this PR. 
In my (current) opinion it is a bug that they ever passed an array to the callback function (the reason being they could be passed to native `array_map` and `array_filter` that way). This change fixes that. `map` and `filter` will get a nice row object as parameter.

This is a breaking change, will update the version accordingly. I've already checked Subsidieportaal for compatibility and we're good to go, after I do 1 minor tweak to a `map` implementation.

File: library/Garp/Db/Table/Rowset.php [hunk 1]
```diff
@@ -49,37 +79,59 @@ public function flatten($column) {
     /**
      * Project a rowset using $fn
      *
-     * @param  callable $fn
+     * @param  callable $transform
      * @return Garp_Db_Table_Rowset
      */
```
Comment (martijngastkemper) [inline library/Garp/Db/Table/Rowset.php]: Due to the transforming to an array virtual field will get lost. I think it's good to mention that.

File: library/Garp/Db/Table/Rowset.php [hunk 2]
```diff
@@ -49,37 +79,59 @@ public function flatten($column) {
...
-    public function map(callable $fn) {
-        $rows = array_map($fn, $this->toArray());
-        $out = new static([
+    public function map(callable $transform): Garp_Db_Table_Rowset {
+        $mapped = $this->reduce(
+            function (array $acc, Zend_Db_Table_Row_Abstract $row) use ($transform): array {
```
Comment (martijngastkemper) [inline library/Garp/Db/Table/Rowset.php]: Will you add this breaking change to `CHANGELOG.md`?

**Comment by harmenjanssen** [comment]
Changelog is bijgewerkt. 
Wat vind je van die Virtual oplossing?

**Pull request merged**